### ISSUE TITLE
SQL server hangs with more than one connection

### ISSUE THREAD

**Issue by wegoodwin** [denisenkom/go-mssqldb#472]
I am using sql server 14.  I have 2 databases on my local machine.  I am trying to copy data from one of the databases to the other.  However after the second connection when I try to do a simple select the program just hangs.  Below is some sample code:

```
sourceconn, err := sql.Open("mssql", "server=localhost;user id=?;password=?;database=dbsource;Trusted_Connection=no;connection timeout=30;")
	if err != nil {
		log.Println("copyData()Err001")
		log.Fatal(err)
	}

	targetconn, err := sql.Open("mssql", "server=localhost;user id=?;password=?;database=dbtarget;Trusted_Connection=no;connection timeout=30;")

	if err != nil {
		log.Println("copyData()Err002")
		log.Fatal(err)
	}

        var tablectr int64

//hangs right here
	err := sourceconn.QueryRow("select count(*) from table1").Scan(&tablectr)
	if err != nil {
		log.Println("getTableCount()Err001")
		log.Fatal(err)
	}
```

**Comment by v-kaywon** [comment]
@wegoodwin I tried your sample code but I was not able to reproduce the hang. I tried with both SQL Server 2014 SP2 Enterprise and SQL Server 2014 Express.

One small bug I noticed in your sample app, you are declaring the `err` variable again at line:
`err := sourceconn.QueryRow("select count(*) from table1").Scan(&tablectr)`

I'm not sure if that is the cause of the hang you're experiencing (from my understanding the code should not have compiled since `err` is declared twice), but can you try to fix it to see what happens?

**Comment by wegoodwin** [comment]
Hi @v-kaywon ,

That is a copy/paste error.  That is not the cause of the problem.  If you are able to connect to 2 databases on the same localhost and have no issues then I could assume that it is something wrong with the way sql server is setup on my machine.  I will modify the code so that it will reduce confusion.

**Comment by v-kaywon** [comment]
@wegoodwin  Can you check if you have any active transaction on table1? An active transaction may cause it to hang.

**Comment by wegoodwin** [comment]
Hi @v-kaywon ,

I restarted my machine.  There are no other programs that are connected to the Database.  


I ran the following sql:

`SELECT * FROM sys.sysprocesses WHERE open_tran = 1;`

and I have attached the results.  I am not sure what this means.


[result.txt](https://github.com/denisenkom/go-mssqldb/files/3155075/result.txt)

**Comment by v-kaywon** [comment]
Yes that means you do have an active transaction from the process with spid 55. Can you check why that process is needed? [Here](https://www.mssqltips.com/sqlservertip/2732/different-techniques-to-identify-blocking-in-sql-server/) is a good post for detecting blocking in SQL Server.

**Comment by wegoodwin** [comment]
Hi @v-kaywon ,

I think I have figured out the problem, thanks to your help.  The wait type I got was LCK_M_SCH_S.  I am doing an alter table to remove all constraints from the target table.  If I comment that code out SQL Server no longer hangs.  I will do more research on this issue.  Thanks for your help.  I will also include the offending lines of code below so that it will help the next person who runs into this issue.



```
sourceconn, err := sql.Open("mssql", "server=localhost;user id=?;password=?;database=dbsource;Trusted_Connection=no;connection timeout=30;")
	if err != nil {
		log.Println("copyData()Err001")
		log.Fatal(err)
	}

	targetconn, err := sql.Open("mssql", "server=localhost;user id=?;password=?;database=dbtarget;Trusted_Connection=no;connection timeout=30;")

	if err != nil {
		log.Println("copyData()Err002")
		log.Fatal(err)
	}

	tx, terr := targetconn.Begin()
		if terr != nil {
			log.Println("copyData()Err003")
			log.Fatalln(terr)
		}

//These lines need to be commented out or SQL Server will hang.
	tx.Exec("ALTER TABLE " + rSQLs[r].schemaname + "." + rSQLs[r].tablename + " NOCHECK CONSTRAINT ALL")
			tx.Exec("SET IDENTITY_INSERT " + rSQLs[r].schemaname + "." + rSQLs[r].tablename + " ON")
        var tablectr int64

//hangs right here
	err2 := sourceconn.QueryRow("select count(*) from table1").Scan(&tablectr)
	if err2 != nil {
		log.Println("getTableCount()Err001")
		log.Fatal(err2)
	}


```
